src/Bicep.Core/Emit/EmitLimitationCalculator.cs

```diff
@@ -351,33 +353,82 @@ public static void DetectInvalidValueForParentProperty(SemanticModel semanticMod
+                        continue;
+                    }
+
+                    // lambdas are not valid inside any other syntax - raise an error and exit
+                    diagnostics.Write(lambda, x => x.LambdaFunctionsOnlyValidInFunctionArguments());
+                    break;
+                }
+            }
+        }
+
+        private static void BlockUnsupportedLambdaVariableUsage(SemanticModel model, IDiagnosticWriter diagnostics)
+        {
+            IEnumerable<LocalVariableSymbol> CollectLambdaVariables(SyntaxBase syntax)
+            {
+                return SyntaxAggregator.AggregateByType<VariableAccessSyntax>(syntax)
+                    .Select(v => model.Binder.GetSymbolInfo(v))
+                    .OfType<LocalVariableSymbol>()
+                    .Distinct()
+                    .Where(symbol => symbol.LocalKind == LocalKind.LambdaItemVariable);
+            }
+
+            var visited = new HashSet<SyntaxBase>();
+            foreach (var lambda in SyntaxAggregator.AggregateByType<LambdaSyntax>(model.Root.Syntax))
+            {
+                foreach (var functionCall in SyntaxAggregator.AggregateByType<FunctionCallSyntaxBase>(lambda.Body))
                 {
-                    var parent = model.Binder.GetParent(lambdaSyntax);
-                    while (parent is not null)
+                    // Block the usage of lambdas inside reference() or list*() functions.
+                    // The Deployment Engine needs to be able to process these upfront to build the deployment graph, so they cannot contain unevaluated lambda variables.
+                    if (!visited.Contains(functionCall) &&
+                        model.GetSymbolInfo(functionCall) is FunctionSymbol functionSymbol &&
+                        functionSymbol.FunctionFlags.HasFlag(FunctionFlags.RequiresInlining))
                     {
-                        if (parent is FunctionArgumentSyntax)
+                        var blockedSymbols = functionCall.Arguments.SelectMany(x => CollectLambdaVariables(x.Expression)).Distinct()
+                            .Select(s => s.Name).ToImmutableArray();
+
+                        if (blockedSymbols.Any())
                         {
-                            // we're inside a function argument - all good
-                            return true;
+                            diagnostics.Write(functionCall, x => x.LambdaVariablesInInlineFunctionUnsupported(functionCall.Name.IdentifierName, blockedSymbols));
                         }
+                    }
+
+                    visited.Add(functionCall);
+                }
 
-                        if (parent is ParenthesizedExpressionSyntax)
+                foreach (var arrayAccess in SyntaxAggregator.AggregateByType<ArrayAccessSyntax>(lambda.Body))
+                {
+                    // Block the usage of lambdas to index into arrays or resources, as this may result in the generation of a reference() or list*() function call.
```

Comment:
```suggestion
                    // Block the usage of lambdas to index into arrays of resources, as this may result in the generation of a reference() or list*() function call.
```